src/steps/resource-manager/monitor/client.ts

```diff
@@ -94,4 +96,30 @@ export class MonitorClient extends Client {
       callback,
     });
   }
+
+  public async iterateActivityLogs(
+    resourceId: string,
+    callback: (e: EventData) => void | Promise<void>,
+    dayRange?: number,
+  ) {
+    const serviceClient = await this.getAuthenticatedServiceClient(
+      MonitorManagementClient,
+    );
+
+    const startDate = formatISO(subDays(new Date(), dayRange || 90));
+    const endDate = formatISO(new Date());
+
+    return iterateAllResources({
+      logger: this.logger,
+      serviceClient,
+      resourceEndpoint: {
+        list: async () =>
+          serviceClient.activityLogs.list(
+            `eventTimestamp ge '${startDate}' and eventTimestamp le '${endDate}' and resourceUri eq '${resourceId}'`,
+          ),
+      },
+      resourceDescription: 'monitor.activityLogs',
+      callback,
+    });
+  }
```

Comment:
This endpoint has both a `list` and `listNext` method. You'll need to implement the `listNext` function as well here.

```suggestion
    return iterateAllResources({
      logger: this.logger,
      serviceClient,
      resourceEndpoint: {
        list: async () =>
          serviceClient.activityLogs.list(
            `eventTimestamp ge '${startDate}' and eventTimestamp le '${endDate}' and resourceUri eq '${resourceId}'`,
          ),
          listNext: async (nextPageLink) => serviceClient.activityLogs.listNext(nextPageLink),
      },
      resourceDescription: 'monitor.activityLogs',
      callback,
    });
  }
```

Reply:
Yep, saw it too. I'll include it in the incoming changes 👍. Thanks!

Reply:
# Additional questions

## What are the required permissions to view Azure Audit logs?

In the [documentation](https://docs.microsoft.com/en-us/azure/active-directory/reports-monitoring/concept-audit-logs#who-can-access-it):
> To access the audit logs, you need to be in one of the following roles:
> - Security Administrator
> - Security Reader
> - Report Reader
> - Global Reader
> - Global Administrator

We've also ran the related test using an account that has similar permissions to JupiterOne users. The test passed which means we don't need to add extra permissions. This is also the expected behavior since an account with a Global Reader role has access to activity logs as stated above.

## How will the activity logs API affect execution times and API rate limits?

### Execution time

There's little to no documentation about any analysis of the activity logs endpoint but it is safe to say that the execution time of this integration step will increase directly proportional to the amount of storage accounts the user has and its activity. We can only optimize the number of iterations by querying the activity logs of the specific resource - in this case, storage account - as seen in the [`$filter`](https://github.com/Creativice-Oy/graph-azure/blob/552c847c8fad7fd1711baf700c5b039e09af59da/src/steps/resource-manager/monitor/client.ts#:~:text=resourceUri%20eq%20%27%24%7BresourceId%7D%27) argument of the endpoint used. Doing this will limit the scope of the activity logs that are called specifically to the storage account resource that is being integrated. I've also utilized the [`$select`](https://github.com/Creativice-Oy/graph-azure/blob/552c847c8fad7fd1711baf700c5b039e09af59da/src/steps/resource-manager/storage/index.ts#:~:text=select%3A%20%27authorization%2C%20status%2C%20eventTimestamp%27%2C) option of the endpoint so we only receive the fields that are needed. This is to optimize the memory used by the integration.

There also seems to be a maximum of 5000 activity logs to be queried as [seen in this part of the documentation](https://docs.microsoft.com/en-us/azure/active-directory/reports-monitoring/reports-faq#how-many-records-i-can-download-from-azure-portal-).

### API limits

The activity logs endpoint is called per storage account which means that the number of API request that is added to comply with this benchmark is equal to the number of storage accounts. There is no way around this as we really need to query for the activity logs per storage account based on the benchmark documentation and it doesn't seem to be excessive as this pattern can be seen all around the Azure integration.

Cc. @ndowmon